src/cmd/troubleshoot/builder.go

```diff
@@ -88,27 +92,89 @@ func (builder CommandBuilder) buildRun() func(*cobra.Command, []string) error {
 		}
 
 		apiReader := k8scluster.GetAPIReader()
-		tests := []troubleshootFunc{
+		dynakubeUnrelatedTests := []troubleshootFunc{
 			checkNamespace,
-			checkDynakube,
-			checkDTClusterConnection,
-			checkImagePullable,
 		}
+
 		troubleshootCtx := troubleshootContext{
 			context:       context.Background(),
 			apiReader:     apiReader,
 			httpClient:    &http.Client{},
 			namespaceName: namespaceFlagValue,
-			dynakubeName:  dynakubeFlagValue,
 		}
-
-		for _, test := range tests {
+		for _, test := range dynakubeUnrelatedTests {
 			err = test(&troubleshootCtx)
 			if err != nil {
 				logErrorf(err.Error())
 				return nil
 			}
 		}
+		resetLog()
+
+		perDynakubeTests := []troubleshootFunc{
+			checkDynakube,
+			checkDTClusterConnection,
+			checkImagePullable,
+		}
+
+		troubleshootCtx.dynakubeName = dynakubeFlagValue
+		dynakubes, err := getDynakubes(troubleshootCtx)
+		if err != nil {
+			return nil
+		}
+
+		for _, dynakube := range dynakubes {
+			troubleshootCtx = troubleshootContext{
+				context:       context.Background(),
+				apiReader:     apiReader,
+				httpClient:    &http.Client{},
+				namespaceName: namespaceFlagValue,
+				dynakube:      dynakube,
+				dynakubeName:  dynakube.Name,
+			}
+			logNewDynakubef(troubleshootCtx.dynakubeName)
+			for _, test := range perDynakubeTests {
+				err = test(&troubleshootCtx)
+				if err != nil {
+					logErrorf(err.Error())
+					return nil
+				}
+			}
+			resetLog()
+			logOkf("'%s' - all checks passed", troubleshootCtx.dynakubeName)
+		}
 		return nil
 	}
 }
+
+func getDynakubes(troubleshootCtx troubleshootContext) ([]dynatracev1beta1.DynaKube, error) {
+	var err error
+	var dynakubes []dynatracev1beta1.DynaKube
+	if troubleshootCtx.dynakubeName == "" {
+		logNewDynakubef("no Dynakube specified - checking all Dynakubes in namespace '%s'", namespaceFlagValue)
+		dynakubes, err = getAllDynakubesInNamespace(troubleshootCtx)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		dynakube := dynatracev1beta1.DynaKube{}
+		dynakube.Name = troubleshootCtx.dynakubeName
+		dynakubes = append(dynakubes, dynakube)
```

Comment:
the actual dynakube could be queried and used here since the list operation also queries the actual ones

Reply:
Will do

Reply:
Do not agree - the list operation will not fail as we do not request a specific dynakube. 

In the case of querying one single dynakube it could happen that the dynakube does not exists, which is checked in the dynakube subtests

Reply:
True, I'd still pass the dynakubeName flag as a parameter instead of having it in the context so the field can be removed

Reply:
Already moved 👍